libs/shared/components/src/Inputs/BigintInput.tsx

```diff
@@ -0,0 +1,106 @@
+import { forwardRef, useEffect, useState } from 'react';
+
+import { InputBase } from '@mui/material';
+import { isNilOrEmpty } from '@origin/shared/utils';
+import { formatUnits, parseUnits } from 'viem';
+
+import type { InputBaseProps } from '@mui/material';
+import type { ChangeEvent } from 'react';
+
+export type BigintInputProps = {
+  value: bigint;
+  decimals?: number;
+  onChange?: (value: bigint) => void;
+  isLoading?: boolean;
+  isError?: boolean;
+} & Omit<InputBaseProps, 'value' | 'onChange' | 'inputRef'>;
+
+export const BigintInput = forwardRef<HTMLInputElement, BigintInputProps>(
+  ({ value, decimals = 18, isLoading, isError, onChange, ...rest }, ref) => {
+    const [strVal, setStrVal] = useState(formatUnits(value, decimals));
+
+    useEffect(() => {
+      if (value === 0n && (isNilOrEmpty(strVal) || strVal === '0.')) {
+        return;
+      }
+
+      if (value === 0n && !/0\.0+$/.test(strVal)) {
+        setStrVal('');
+        return;
+      }
+
+      if (
+        isNilOrEmpty(strVal) ||
+        strVal === '0.' ||
+        value !== parseUnits(strVal, decimals)
+      ) {
+        setStrVal(formatUnits(value, decimals));
+      }
+    }, [value, decimals, strVal]);
+
+    const handleChange = (evt: ChangeEvent<HTMLInputElement>) => {
+      if (evt.target.validity.valid) {
+        const val =
+          isNilOrEmpty(evt.target.value) || evt.target.value === '.'
+            ? '0'
+            : evt.target.value.replace(/\.0+$/, '');
+
+        try {
+          const num = parseUnits(val, decimals);
+          setStrVal(evt.target.value === '.' ? '0.' : evt.target.value);
+          if (onChange && num !== value) {
+            onChange(num);
+          }
+        } catch {}
```

Comment:
yeah, this is something to discuss with Matt, I used to change color to `palette.error.main` but it could also be the focus frame... anyway, it's there for now, to be refined